main.cpp

```diff
@@ -0,0 +1,327 @@
+void manageWaterLevel();
+void availableCommands();
+void debounceButton(Button*);
+void buttonsInit();
+void uartTask();
+void manualLevelManagement();
+void checkManualManagementSwitch();
+
+int main()
+{
+    inputsInit();
+    outputsInit();
+    buttonsInit();
+    while(true) {
+        if(manualMode == true) {
+            manualLevelManagement();
+        } else {
+            checkManualManagementSwitch();
+            getTanksWaterLevel();
+            manageWaterLevel();
+        }
+        uartTask();
+    }
+}
+
+void inputsInit() {
+}
+
+/**
+* @brief Iterates over the tanks array and initiate the actuators and leds with the OFF value.
+*/
+void outputsInit() {
+    for (int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].actuator = OFF;
+        tanks[i].led = OFF;
+    }
+}
+
+/**
+* @brief Function used to calculate the water level in the tanks, saving the information in the structs of the tanks array.
+*
+* The sensors readings are saved into the levelSensorsReadingsArray. All the readings in the array are then averaged to
+* obtain a more reliable analog measure.
+*
+* The arrays are filled replacing the older measures by the new ones. At the begining, when the array is empty, the measure
+* is not accurate. In our case, the system gets stable once the array is filled (imperceptible time).
+*
+* The Tank struct and the tanks array helps to avoid code repetition here. All the measures are saved in the Tank structs
+* initialized in the tanks array.
+*/
+void getTanksWaterLevel() {
+    static int levelSensorsSampleIndex = 0;
+
+    for (int i = 0; i < NUMBER_OF_TANKS; i++){
+        tanks[i].levelSensorReadingsArray[levelSensorsSampleIndex] = tanks[i].levelSensor.read();
+    }
+
+    levelSensorsSampleIndex += 1;
+
+    if (levelSensorsSampleIndex >= NUMBER_OF_AVG_SAMPLES) {
+        levelSensorsSampleIndex = 0;
+    }
+
+    for(int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].levelSensorReadingsSum = 0;
+    }
+
+    for(int i = 0; i < NUMBER_OF_TANKS; i++){
+        for (int j = 0; j < NUMBER_OF_AVG_SAMPLES; j++) {
+            tanks[i].levelSensorReadingsSum += tanks[i].levelSensorReadingsArray[j];
+        }
+    }
+    
+    for(int i = 0; i < NUMBER_OF_TANKS; i++) {
+        tanks[i].levelSensorReadingsAverage = tanks[i].levelSensorReadingsSum / NUMBER_OF_AVG_SAMPLES;
+    }
+}
+
+/**
+* @brief This is the management system for the automatic water level management.
+*
+* Iterate over all tanks in the systems and check if the water level is below the water level threshold.
+* In case it is, turn on the actuator and the led of the tank. In case it is not turn them off.
+*/
+void manageWaterLevel() {
+    for(int i = 0; i < NUMBER_OF_TANKS; i++) {
+        if (tanks[i].levelSensorReadingsAverage < WATER_LEVEL_THRESHOLD) {
+            tanks[i].actuator = ON;
+            tanks[i].led = ON;
+        } else {
+            tanks[i].actuator = OFF;
+            tanks[i].led = OFF;
+        }
+    }
+}
+
```

Comment:
Real life will tell if this oscillates or not.
I bet it will.
As long as water can freely move, if something (filling ?) starts making waves, level will go above and below the threshold. Tinkering with the averaging time might smooth it out, but eventually it will say hi again.

Reply:
@scaprile you're right. Maybe I could improve the system using two thresholds: a lower threshold that starts the filling process and an upper threshold that stops the filling process. I think that properly calibrating these thresholds would solve the issue.

For example, the upper threshold could be 85% of the tank capacity and the lower threshold could be 80% of the tank capacity. In that case, if the tank is being filled and the lower threshold is exceeded, the water pump (or solenoid valves) will keep filling the tank until the upper threshold is reached. If water waves or vibrations reach the upper level, the system will stop the filling process. In that case, the system will not oscillate since the lower threshold will not be reached yet.